alf/experience_replayers/replay_buffer.py

```diff
@@ -253,6 +343,148 @@ def _prioritized_sample(self, batch_size, batch_length):
         indices = self._sum_tree.find_sum_bound(r)
         return self._index_to_env_id_pos(indices)
 
+    def pad(self, x, env_ids):
+        """Make indices in RingBuffer non-circular.
+
+        This is useful when trying to compute distance from one index to the
+        other, and is done by adding _max_length when index is smaller than
+        ``current_ending_pos - current_size``."""
+        return torch.where(
+            x < self.circular(self._current_pos[env_ids] -
+                              self._current_size[env_ids]),
+            x + self._max_length, x)
+
+    def store_episode_end_pos(self, non_first, pos, env_ids):
+        """Update _index and _recent_overwritten for episode ending pos.
+
+        Args:
+            non_first_idx (tensor): index of the added batch of exp, which are
+                not FIRST steps.  We need to update last step pos for all
+                these env_ids[non_first_idx].
+            pos (tensor): position of the stored batch.
+            env_ids (tensor): env_ids of the stored batch.
+        """
+        _env_ids = env_ids[non_first]
+        _pos = pos[non_first]
+        prev_pos = self.circular(_pos - 1)
+        prev_first = self._index[(_env_ids, prev_pos)]
+        padded_prev_first = self.pad(prev_first, _env_ids)
```

Comment:
If we never apply circular() to position, we don't need to pad.

Reply:
True here, but not always true in general, because pos in _indexed_pos becomes outdated as we add new data.